### PR TITLE
Fix Code Todos

### PR DESCRIPTION
**Pull request by PhilippSiegmantel**

### CODE REVIEW

File: src/ui/components.tsx [hunk 1]
```diff
@@ -365,6 +365,8 @@ export const HostLabelFilter: React.FC<{
 };
 
 export const OnlyActiveChildren = (props: { children: JSX.Element[]; requestSpec: RequestSpec }): JSX.Element => {
+  type ChildComponent = React.ReactElement<{label: string; onChange(value: unknown): void}, JSXElementConstructor<unknown>>
```
Comment (PhilippSiegmantel) [inline src/ui/components.tsx]: Link `onChange` to RequestSpec[T]` where `T` is `RequestSpecKyes` that can be undefined or something along the lines.

File: src/backend/rest.ts [hunk 2]
```diff
@@ -66,7 +66,7 @@ export default class RestApiBackend implements Backend {
   }
 
   async testDatasource(): Promise<unknown> {
-    const result = await this.api<{versions: {checkmk: string}; edition: string}>({
+    const result = await this.api<{ versions: { checkmk: string }; edition: string }>({
```
Comment (BenediktSeidl) [inline src/backend/rest.ts:69]: please remove the "fix formatting" commit and fix the formatting in those commits that break the formatting.

File: tests/cypress/e2e/helpers.ts [hunk 3]
```diff
@@ -1,3 +1,13 @@
+/* TODO:
```
Comment (BenediktSeidl) [inline tests/cypress/e2e/helpers.ts:1]: DRY ;-) i think it would be better to mention the other file in the comment.

File: src/ui/components.tsx [hunk 4]
```diff
@@ -4,16 +4,16 @@ import {
   AsyncSelect,
   Button,
   Checkbox,
-  Select as GrafanaSelect,
   HorizontalGroup,
   InlineField,
   InlineFieldRow,
   Input,
   Label,
+  Select as GrafanaSelect,
```
Comment (BenediktSeidl) [inline src/ui/components.tsx]: looks like one tool sorts by the first character, the other by the character it imports from? if this causes __no__ prettier error in ci, we have to adapt the ci to make this stable.

File: tests/unit/QueryEditor.test.tsx [hunk 5]
```diff
@@ -113,12 +113,12 @@ describe('QueryEditor RAW', () => {
     expect(onRunQuery).toHaveBeenCalledTimes(1);
   });
 
-  it.skip.each`
```
Comment (BenediktSeidl) [inline tests/unit/QueryEditor.test.tsx:116]: typo in the commit message (can not comment on commit message?! )-:

File: src/RequestSpec.ts [hunk 6]
```diff
@@ -36,9 +36,6 @@ export interface TagValue {
   operator?: string;
 }
 
-// TODO: is unused.
```
Comment (BenediktSeidl) [inline src/RequestSpec.ts:39]: could/should we try to use this?
same thing for graph_type: we know this can only be `graph` or `metric`?
perhaps you can give this a try. if too many errors pop up, we can remove it.

Comment (BenediktSeidl) [reply src/RequestSpec.ts:39]: btw: you added this type as "Presentation" in the commit before this commit. perhaps we can rename it here? should we rename "presentation" to aggregation, and use this type instead?

File: src/types.ts [hunk 7]
```diff
@@ -1,6 +1,6 @@
 import { DataQuery, DataSourceJsonData } from '@grafana/data';
 
-import { RequestSpec, defaultRequestSpec } from './RequestSpec';
+import { defaultRequestSpec, RequestSpec } from './RequestSpec';
```
Comment (BenediktSeidl) [inline src/types.ts]: if the ci is ok with this change, we have to change the ci. import ordering should be stabilized.

**Review by BenediktSeidl — CHANGES_REQUESTED** [review]

File: src/ui/components.tsx (hunk 4 above)
Comment (PhilippSiegmantel) [reply src/ui/components.tsx]: Seems to be an error in the ci. See [here](https://github.com/tribe29/grafana-checkmk-datasource/actions/runs/3909674284/jobs/6681008671)

File: src/types.ts (hunk 7 above)
Comment (PhilippSiegmantel) [reply src/types.ts:3]: It's an CI error as well. See [here](https://github.com/tribe29/grafana-checkmk-datasource/actions/runs/3909744017/jobs/6681153601)

File: tests/unit/QueryEditor.test.tsx (hunk 5 above)
Comment (PhilippSiegmantel) [reply tests/unit/QueryEditor.test.tsx:116]: Could you tell me which message you mean? I didn't see one when proofreading just now.

File: src/RequestSpec.ts (hunk 6 above)
Comment (PhilippSiegmantel) [reply src/RequestSpec.ts:39]: The name seems to have been carried over from the WebApi, but Aggregation make a lot more sense to me.

File: tests/unit/QueryEditor.test.tsx (hunk 5 above)
Comment (BenediktSeidl) [reply tests/unit/QueryEditor.test.tsx:116]: https://github.com/tribe29/grafana-checkmk-datasource/pull/126/commits/c636e87834db865a47feaccaf2e81e845a214d6f

> `it is and is`

feels like this was reworded and some words went missing/duplicated?

File: src/ui/components.tsx (hunk 4 above)
Comment (BenediktSeidl) [reply src/ui/components.tsx]: can you explain?

what exactly i the ci error? that the current main branch is green?
how can we fix this? can you create another pr for fixing this?

Comment (PhilippSiegmantel) [reply src/ui/components.tsx]: Oh, what I meant was that this is caught by prettier both in the CI and locally in my tests. The linked action was a test commit where I tried and succeeded to provoke that error in particular.

File: src/ui/components.tsx [hunk 8]
```diff
@@ -155,7 +153,7 @@ export const CheckMkSelectNegatable = <T extends RequestSpecNegatableOptionKeys>
         }
       : { ...props.value };
 
-  const onValueChange = (newValue: string | undefined) => {
+  const onValueChange = (newValue: string) => {
     if (newValue === undefined) {
```
[the remark this answers is not in the document]
Comment (PhilippSiegmantel) [reply src/ui/components.tsx:161]: Looks like a line got lost when I split the change. 

A few lines above, `value.value` is set to an empty string instead of undefined. Fixed now.

File: src/ui/components.tsx [hunk 9]
```diff
@@ -150,7 +154,7 @@ export const CheckMkSelectNegatable = <T extends RequestSpecNegatableOptionKeys>
   const value: NegatableOption =
     props.value === undefined
       ? {
-          value: '',
```
Comment (BenediktSeidl) [inline src/ui/components.tsx:153]: why was this changed? have you tested it?

Comment (PhilippSiegmantel) [reply src/ui/components.tsx:153]: It was the change that got lost during commit splitting I talked about. I tested it and now it displays the placeholder if no value was passed as a prop.

Comment (PhilippSiegmantel) [reply src/ui/components.tsx:153]: I'll do that

File: src/ui/QueryEditor.tsx [hunk 10]
```diff
@@ -102,7 +101,10 @@ export const QueryEditor = (props: Props): JSX.Element => {
 
   const siteAutocompleter = React.useCallback(
     (prefix: string) => {
-      return contextAutocomplete(datasource, 'sites', {}, prefix, { strict: false });
+      // TODO: this is a quick fix until we have decided how to handle site='' in raw edition
```
Comment (BenediktSeidl) [inline src/ui/QueryEditor.tsx:104]: this TODO should be resolved, i think this is a rebase error!

File: src/ui/QueryEditor.tsx [hunk 11]
```diff
@@ -87,7 +86,7 @@ export const QueryEditor = (props: Props): JSX.Element => {
     onRunQuery();
   }
 
-  const presentationCompleter = async (): Promise<Array<SelectableValue<Presentation>>> => [
+  const presentationCompleter = async (): Promise<Array<SelectableValue<Aggregation>>> => [
```
Comment (BenediktSeidl) [inline src/ui/QueryEditor.tsx:89]: i think this is a rebase error and should not be changed.

File: src/utils.ts [hunk 12]
```diff
@@ -1 +1,79 @@
+import { isUndefined } from 'lodash';
+
+import { NegatableOption, RequestSpec } from './RequestSpec';
+import { CmkQuery } from './types';
+import { requestSpecFromLegacy } from './webapi';
+
 export const titleCase = (str: string): string => str[0].toUpperCase() + str.slice(1).toLowerCase();
+
+export function contextEntryFromNegatableOption(option: NegatableOption, keyName: string) {
```
Comment (BenediktSeidl) [inline src/utils.ts:9]: please mention this new function in the commit message, or better: move it to a new commit.
did not expect code changes in this commit.

**Review by BenediktSeidl — CHANGES_REQUESTED** [review]

File: src/ui/QueryEditor.tsx [hunk 13]
```diff
@@ -102,7 +101,8 @@ export const QueryEditor = (props: Props): JSX.Element => {
 
   const siteAutocompleter = React.useCallback(
     (prefix: string) => {
-      return contextAutocomplete(datasource, 'sites', {}, prefix, { strict: false });
+      const strict = datasource.getEdition() === 'RAW';
+      return contextAutocomplete(datasource, 'sites', {}, prefix, { strict: strict });
```
[the remark this answers is not in the document]
Comment (PhilippSiegmantel) [reply src/ui/QueryEditor.tsx]: Got it. Should be fixed now.
